### PR TITLE
Fix reports test to run under Unit-test and other environments

### PR DESCRIPTION
**Pull request by prculley**
Originally ran correctly only under nose; under unit-test failed to find anything to do.  Caused by test method not having 'test' in the name.
Also was selecting wrong test file from example directory instead of tests directory.  Not yet critical, but changes to example could cause tests to fail.
When attempting to debug under my personal environment, couldn't find the test file.  So set up test to use DATA_DIR, instead of the __file__ dir as root, which seems to work in all environments.

Also did pylint improvements on report_test as separate commit.

Found these when trying to debug the tests for the detailed descendant report.

### CODE REVIEW

**Comment by sam-m888** [comment]
Only tested on Linux

Results (note time increase for patched version):

* without patch using unit-test on local pc

> Ran 21676 tests in 391.545s
> 
> OK (skipped=56)
> 


* with patch using unit-test on local pc

> Ran 21676 tests in 403.764s
> 
> OK (skipped=56)
> 


* without patch using nose on travis-ci

> Ran 21888 tests in 556.893s
> 
> OK
> 


* with patch using nose on travis-ci

> Ran 21888 tests in 615.729s
> 
> OK
> 

* without patch running single test on pc (verbose mode)

> GRAMPS_RESOURCES=. python -m unittest gramps.plugins.test.reports_test -v
> 
> <snipped error results etc from here>
> 
> Ran 0 tests in 0.000s
> 
> OK
> 


* with patch running single test on pc (verbose mode)

> GRAMPS_RESOURCES=. python -m unittest gramps.plugins.test.reports_test -v
> 
> <snipped successful test run of each report>
> 
> Ran 30 tests in 15.056s
> 
> OK
>

**Comment by Nick-Hall** [comment]
Good work!  Are all unit tests now run by `python setup.py test`?

**Comment by jralls** [comment]
With the ```nosetests``` command lifted from .travis.yml on this feature branch I get Ran 687 tests in 304.556s: FAILED (SKIP=1, errors=11, failures=7) while on master I get Ran 687 tests in 304.837s: FAILED (SKIP=1, errors=22, failures=6).

With ```python setup.py test``` I get Ran 475 tests in 263.318s with 10 Es and 1 s on both this feature branch and on master.

**Comment by prculley** [comment]
I'm trying to figure some of this out now.  It seems that running unittest on the whole thing with discovery is still not finding any of the tests in the gramps/plugins/test directory and possibly others.

We should be getting the 21888 tests run number.  I finally figured out how to get a verbose listing (verbose=2) with unittest which includes the test names.  Going to try the same for nose next.

I'm pretty sure that variations in execution time under travis are due to server load, I've looked at a bunch of recent commit test runs and find a lot of variation in time, even though the code tested is only varying a little bit.

**Comment by jralls** [comment]
So what ```unittest``` and ```nosetest``` commands are you using to get so many more tests?

**Comment by prculley** [comment]
It seems that unittest started with
`python3 -m setup.py test` or `GRAMPS_RESOURCES=. python3 -m unittest discover -v -p "*_test.py"`
don't find all the tests.

I think the reason is that we don't have an "__init__" file in many of the test directories.  In the case of the plugins, we also don't have that file in the "gramps/plugins" directory, probably deliberately.  This seems to mean (if I understand the docs correctly) that all of these test files are not part of the 'package', and have to be listed separately for testing.

We can either change the setup.py script to explicitly include the missing tests (I think this makes sense for plugins directory) or add in the missing "__init__" files (which I think is reasonable for the other tests).

We may also have to fix up setup.py to skip some tests if we include everything, just like the Travis nosetest script does.

Clearly the discovery that nose uses is more forgiving and is able to find more tests...

**Comment by Paul-Franklin** [comment]
When adding an __init__.py it can be zero-length.
No GPL boilerplate (or anything else) is needed.

**Comment by Nick-Hall** [comment]
@jralls Are you getting sql transaction errors?  If so, try commenting out the "begin" and "commit" lines introducted into the _set_metadata method in commit 0813ba25dbf7e287fe16e8385da7b8181455eabb.

When I do this I get:
```
Ran 21676 tests in 951.311s
OK (skipped=1)

```

**Comment by jralls** [comment]
Reverting 0813ba2 gets me a lot further: Ran 11342 tests in 371.694s, FAILED (SKIP=1, errors=124, failures=7)

I see a bunch of errors about a database that needs recovery and a few where I'm missing modules.
I'm off to the National Genealogical Society conference first thing tomorrow morning so tracking all of that down will have to wait a week.

**Comment by Nick-Hall** [comment]
Added missing `__init__.py` in commit 5ffb4bcfac7a1d431964da49ef850f51675c27f5.

The tests now run, but I get two failures and a large amount of extaneous output.

**Comment by prculley** [comment]
I'm still trying to get the two test methods to produce the same results.
Nosetest using the Travis script lines, and 'setup test'.  I've found several issues, (including probably the same __init__.py files @Nick-Hall added in his commit, but keep finding more as I go along.

I'm trying to compare the tests run by each method line by line, to see where differences occur; I've gotten pretty far but some work yet to do.  There were more substantive differences than I expected...  As well as difference caused by the way the two test methods work and report results.

To enable the comparison I patched up the imports_test, so that nose output didn't just repeat the docstring, also patched it so that unittest did not run those tests twice!!!

Just  FYI in case anyone else wants to avoid duplicating my work (or vice versa).  I'll report back with some more commits when I have things better in hand.

**Comment by prculley** [comment]
Between these changes and what @Nick-Hall did on master, I think the unittest and nosetest are almost the same.  Run the lines from Travis script for nosetest, and `python3 -m setup.py test` for unittest.

There are still some differences, but I will leave these for another bug/PR etc.

On my system, nosetest finds the following and unittest does not:
todo_test_replace_child (gramps.gen.lib.test.merge_test.CitationBaseCheck) ... ok
todo_test_replace_child (gramps.gen.lib.test.merge_test.NoteBaseCheck) ... ok
todo_test_merge_person_aschild (gramps.gen.lib.test.merge_test.PersonCheck) ... ok
todo_test_merge_person_asparent (gramps.gen.lib.test.merge_test.PersonCheck) ... ok
todo_test_replace (gramps.gen.lib.test.merge_test.SourceCheck) ... ok
gramps.gui.logger.test.rotate_handler_test.testSuite ... ok
gramps.plugins.db.bsddb.test.cursor_test.testSuite ... ok
gramps.plugins.db.bsddb.test.reference_map_test.testSuite ... ok
Test that simple recall of messages works. ... info message ok
ErrorReportAssistant_Test.testSuite ... ok
Test that the window appears. ... An info message A warn message A debug message ok
GtkHandler_Test.testSuite ... ok

File: setup.py [hunk 1]
```diff
@@ -300,7 +300,8 @@ def run(self):
             raise RuntimeError("No build directory. Run `python setup.py build` before trying to run tests.")
         os.environ['GRAMPS_RESOURCES'] = '.'
         all_tests = unittest.TestLoader().discover('.', pattern='*_test.py')
-        unittest.TextTestRunner(verbosity=self.verbose).run(all_tests)
+        unittest.TextTestRunner(
+            verbosity=2 if self.verbose else 0).run(all_tests)
```
Comment (Nick-Hall) [inline setup.py]: This causes the tests to be run in verbose mode even if the `-v` option is not specified.

**Comment by Nick-Hall** [comment]
I now see 21877 tests run in both Travis CI and `python setup.py test`.  I have created bug reports for the three remaining issues:

10038: Unit tests report 108 errors
10039: Unit tests report 2 failures
10040: Extraneous output in unit tests

File: setup.py (hunk 1 above)
Comment (prculley) [inline setup.py]: The distutils package (on which our setup code is based) states that -v (verbose) is the default; -q (quiet) makes things quieter.
What we had before for verbose, wasn't.

**Comment by prculley** [comment]
@Nick-Hall I did not have any errors or failures under Ubuntu Linux for Unit test.  What kind of Linux is Travis based on?
Hope I don't need to set up another VM...
P.S. I have noticed that sometimes errors occur when run in environment with non-default settings of one type or another.  I spent part of a day debugging why the some of the reports_test were failing, until I realized that I had set my default paper orientation to Landscape for reports.  I've had other similar cases.

**Comment by Nick-Hall** [comment]
Travis CI is using Ubuntu 14.04.  I get the errors and failures running under Arch Linux as stated in my bug reports.  The extaneous output can be seen under both Ubuntu and Arch.

File: setup.py (hunk 1 above)
Comment (prculley) [inline setup.py]: That's weird; I was running via `python3 -m setup.py test -v` and I only got a row of dots.
`python3 -m setup.py test -q` made it go quiet.  So I thought that the `-v` was having no extra effect, reinforced by the distutils docs.
I now see that `python3 -m setup.py -v test` with '-v' before the 'test' has the desired effect.  The '-q' seems to work in either position...  I'll remove that part of the PR.

**Comment by Nick-Hall** [comment]
@prculley Suppressing the extraneous output is just a matter of commenting out print statements and setting the threshold for log messages as you suggest.  The warnings appear to be from just two plugins.  I'll fix this myself, but thanks for offering to help.

**Comment by wroldwiedbwe** [comment]
> extraneous output

@Nick-Hall @prculley  Checked and the latest gramp50 unittest have extraneous output again.

> $ python setup.py test
> running test
> .................................................................................................................................................test_Complex_1: 3.02
> 
> .test_HasAddress: 0.33
> 
> .test_HasAssociation: 0.39
> 
> .test_HasBirth: 0.73
> 
> .test_HasCommonAncestorWithFilterMatch: 1.99
> 
> .test_HasDeath: 0.64
> 
> .test_HasEvent: 0.80
> 
> .test_HasFamilyAttribute: 0.61
> 
> .test_HasFamilyEvent: 1.10
> 
> .test_HasLDS: 0.37
> 
> .test_HasNameOriginType: 0.39
> 
> .test_HasNameType: 0.36
> 
> .test_HasRelationship: 0.73
> 
> .test_HasTextMatchingRegexpOf: 2.28
> 
> .test_HavePhotos: 0.35
> 
> .test_IsChildOfFilterMatch: 1.10
> 
> .test_IsDescendantFamilyOf: 0.40
> 
> .test_IsDescendantFamilyOfFilterMatch: 1.15
> 
> .test_IsDescendantOf: 0.36
> 
> .test_IsDescendantOfFilterMatch: 1.07
> 
> .test_IsLessThanNthGenerationAncestorOfBookmarked: 0.34
> 
> .test_IsLessThanNthGenerationAncestorOfDefaultPerson: 0.33
> 
> .test_IsMoreThanNthGenerationAncestorOf: 0.47
> 
> .test_IsMoreThanNthGenerationDescendantOf: 0.37
> 
> .test_IsParentOfFilterMatch: 1.07
> 
> .test_IsSiblingOfFilterMatch: 1.11
> 
> .test_IsSpouseOfFilterMatch: 1.17
> 
> .test_IsWitness: 0.38
> 
> .test_ProbablyAlive: 1.86
> 
> .test_RegExpName: 0.51
> 
> .test_commonancestor_empty: 1.24
> 
> .test_commonancestor_irregular: 1.33
> 
> .test_commonancestor_matching: 1.36
> 
> .test_commonancestor_nonmatching: 1.26
> 
> .test_disconnected: 0.37
> 
> .test_dupancestor_empty: 0.37
> 
> .test_dupancestor_irregular: 0.46
> 
> .test_dupancestor_matching: 0.46
> 
> .test_dupancestor_nonmatching: 0.41
> 
> .test_everyone: 0.40
> 
> .test_familywithincompleteevent: 0.86
> 
> .test_hasalternatename: 0.35
> 
> .test_hasidof_empty: 0.37
> 
> .test_hasidof_irregular: 0.38
> 
> .test_hasidof_matching: 0.35
> 
> .test_hasidof_nonmatching: 0.37
> 
> .test_hasnameof: 0.35
> 
> .test_hasnickname: 0.37
> 
> .test_hassoundexname: 0.63
> 
> .test_hassourceof_empty: 0.35
> 
> .test_hassourceof_irregular: 0.33
> 
> .test_hassourceof_matching: 0.56
> 
> .test_hassourceof_nonmatching: 0.33
> 
> .test_hasunknowngender: 0.37
> 
> .test_havealtfamilies: 0.75
> 
> .test_havechildren: 0.74
> 
> .test_incompletenames: 0.36
> 
> .test_isancestoroffiltermatch: 1.41
> 
> .test_isbookmarked: 0.36
> 
> .test_isdefaultperson: 0.35
> 
> .test_isfemale: 0.33
> 
> .test_ismale: 0.37
> 
> .test_isrelatedwith_empty: 0.34
> 
> .test_isrelatedwith_irregular: 0.44
> 
> .test_isrelatedwith_matching: 0.51
> 
> .test_isrelatedwith_nonmatching: 0.35
> 
> .test_missingparent: 0.74
> 
> .test_multiplemarriages: 0.33
> 
> .test_nevermarried: 0.34
> 
> .test_nobirthdate: 0.50
> 
> .test_nodeathdate: 0.43
> 
> .test_peopleprivate: 0.40
> 
> .test_peoplepublic: 0.36
> 
> .test_personwithincompleteevent: 0.59
> 
> .test_relationshipbookmarks: 0.37
> 
> ...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................('person-add', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('family-add', (['0000000500000005', '0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('person-delete', (['0000000300000003', '0000000400000004'],)),
> ('family-delete', (['0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000100000001', '0000000200000002'],)),
> ('family-update', (['0000000500000005'],)),
> ('note-delete', (['0000000700000007'],)),
> ('note-delete', (['0000000700000007'],)),
> ('person-add', (['0000000400000004', '0000000300000003'],)),
> ('family-add', (['0000000600000006'],)),
> ('person-update', (['0000000200000002', '0000000100000001', '0000000200000002', '0000000100000001'],)),
> ('family-update', (['0000000500000005', '0000000600000006', '0000000600000006'],)),
> ('person-delete', (['0000000400000004', '0000000300000003', '0000000200000002', '0000000100000001'],)),
> ('family-delete', (['0000000600000006', '0000000500000005'],)),
> ('person-add', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('family-add', (['0000000500000005', '0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('person-delete', (['0000000300000003', '0000000400000004'],)),
> ('family-delete', (['0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000100000001', '0000000200000002'],)),
> ('family-update', (['0000000500000005'],)),
> ('note-delete', (['0000000700000007'],)),
> .('person-add', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('family-add', (['0000000500000005', '0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('person-delete', (['0000000300000003', '0000000400000004'],)),
> ('family-delete', (['0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000100000001', '0000000200000002'],)),
> ('family-update', (['0000000500000005'],)),
> ('note-delete', (['0000000700000007'],)),
> ('note-delete', (['0000000700000007'],)),
> ('person-add', (['0000000400000004', '0000000300000003'],)),
> ('family-add', (['0000000600000006'],)),
> ('person-update', (['0000000200000002', '0000000100000001', '0000000200000002', '0000000100000001'],)),
> ('family-update', (['0000000500000005', '0000000600000006', '0000000600000006'],)),
> ('person-delete', (['0000000400000004', '0000000300000003', '0000000200000002', '0000000100000001'],)),
> ('family-delete', (['0000000600000006', '0000000500000005'],)),
> ('person-add', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('family-add', (['0000000500000005', '0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000300000003', '0000000400000004'],)),
> ('person-delete', (['0000000300000003', '0000000400000004'],)),
> ('family-delete', (['0000000600000006'],)),
> ('person-update', (['0000000100000001', '0000000200000002', '0000000100000001', '0000000200000002'],)),
> ('family-update', (['0000000500000005'],)),
> ('note-delete', (['0000000700000007'],)),
> ..................................................................................
> ----------------------------------------------------------------------
> Ran 21918 tests in 926.735s
> 
> OK (skipped=0)
>

**Pull request merged**